substrate/runtime/contract/src/vm.rs

```diff
@@ -0,0 +1,769 @@
+	/// transfered from this to the newly created account.
+	fn create(
+		&mut self,
+		code: &[u8],
+		value: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		data: Vec<u8>,
+	) -> Result<CreateReceipt<T>, ()>;
+
+	/// Call (possibly transfering some amount of funds) into the specified account.
+	fn call(
+		&mut self,
+		to: &T::AccountId,
+		value: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		data: Vec<u8>,
+	) -> Result<CallReceipt, ()>;
+}
+
+/// Error that can occur while preparing or executing wasm smart-contract.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+	/// Error happened while serializing the module.
+	Serialization,
+
+	/// Error happened while deserializing the module.
+	Deserialization,
+
+	/// Internal memory declaration has been found in the module.
+	InternalMemoryDeclared,
+
+	/// Gas instrumentation failed.
+	///
+	/// This most likely indicates the module isn't valid.
+	GasInstrumentation,
+
+	/// Stack instrumentation failed.
+	///
+	/// This  most likely indicates the module isn't valid.
+	StackHeightInstrumentation,
+
+	/// Error happened during invocation of the contract's entrypoint.
+	///
+	/// Most likely because of trap.
+	Invoke,
+
+	/// Error happened during instantiation.
+	///
+	/// This might indicate that `start` function trapped, or module isn't
+	/// instantiable and/or unlinkable.
+	Instantiate,
+
+	/// Memory creation error.
+	///
+	/// This might happen when the memory import has invalid descriptor or
+	/// requested too much resources.
+	Memory,
+}
+
+/// Enumerates all possible *special* trap conditions.
+///
+/// In this runtime traps used not only for signaling about errors but also
+/// to just terminate quickly in some cases.
+enum SpecialTrap {
+	/// Signals that trap was generated in response to call `ext_return` host function.
+	Return(Vec<u8>),
+}
+
+struct Runtime<'a, T: Trait + 'a, E: Ext<T> + 'a> {
+	ext: &'a mut E,
+	config: &'a Config<T>,
+	memory: sandbox::Memory,
+	gas_meter: &'a mut GasMeter<T>,
+	special_trap: Option<SpecialTrap>,
+}
+impl<'a, T: Trait, E: Ext<T> + 'a> Runtime<'a, T, E> {
+	fn memory(&self) -> &sandbox::Memory {
+		&self.memory
+	}
+	/// Save a data buffer as a result of the execution.
+	///
+	/// This function also charges gas for the returning.
+	///
+	/// Returns `Err` if there is not enough gas.
+	fn store_return_data(&mut self, data: Vec<u8>) -> Result<(), ()> {
+		let data_len = <T::Gas as As<u64>>::sa(data.len() as u64);
+		let price = (self.config.return_data_per_byte_cost)
+			.checked_mul(&data_len)
+			.ok_or_else(|| ())?;
+
+		match self.gas_meter.charge(price) {
+			GasMeterResult::Proceed => {
+				self.special_trap = Some(SpecialTrap::Return(data));
+				Ok(())
+			}
+			GasMeterResult::OutOfGas => Err(()),
+		}
+	}
+}
+
+fn to_execution_result<T: Trait, E: Ext<T>>(
+	runtime: Runtime<T, E>,
+	run_err: Option<sandbox::Error>,
+) -> Result<ExecutionResult, Error> {
+	let mut return_data = Vec::new();
```

Comment:
Just a small grumble -- we can change this to:

```
let return_data = match (run_err, runtime.special_trap) {
  (None, None) => Vec::new(),
  (Some(sandbox::Error::Execution), Some(SpecialTrap::Return(rd))) => rd,
  ...
}
```

This avoids the `mut` and an extra stack item.